### PR TITLE
Add ja translation for double faults

### LINKED ISSUE #915: Japanese translation of chapters 6 (Double Fault)
I plan to work on Japanese translation of chapter 6 (Double Fault).

Please let me know if anyone is interested in this so that so you don't do any duplicate.

#906

### PR DESCRIPTION
**Pull request by garasubo**
Resolve #915 

Japanese translation for "Double Faults" section

This is my first time PR, so feel free to point out if I'm in the wrong way.

### CODE REVIEW

**Comment by phil-opp** [comment]
Thanks a lot for creating the translation! I'm not speaking Japanese myself, but maybe @JohnTitor or @woodyZootopia could take a look?

File: blog/content/edition-2/posts/06-double-faults/index.ja.md [hunk 1]
```diff
@@ -0,0 +1,556 @@
++++
+title = "Double Faults"
```
Comment (woodyZootopia) [inline blog/content/edition-2/posts/06-double-faults/index.ja.md:2]:
```suggestion
title = "ダブルフォルト"
```

**Comment by woodyZootopia** [comment]
Thank you very much! I think it's really good. There are only a few advice:

* JohnTitor is added as co-translator since he always corrects my translation. You need not add him as for now. (I think he'll help you as well though!)
* I think it's not usual to use _italic characters （イタリック体）_ to emphasize in Japanese. I'd prefer converting it **bold（太字）**.
* It would be more kind to change URLs to the websites in English to the ones in Japanese when translation is available (for example, Wikipedia 日本語版 or Rust by Example 日本語版).

There are other minor mistakes, but it can be corrected after all translation is done. Keep going!

**Comment by garasubo** [comment]
All sections are translated. 

TODO: Replace URLs for English website with those for Japanese websites if exist

**Comment by garasubo** [comment]
Replaced wikipedia links and Rust official document links to those for Japanese. Confirmed that all CI checks passed.
@JohnTitor @woodyZootopia Could you review this PR?

**Comment by woodyZootopia** [comment]
Sorry for the delay in response. I've been really busy with my jobs until next week or so, so it's difficult to review your translation for now. Very sorry.￼￼

File: blog/content/edition-2/posts/06-double-faults/index.ja.md [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+#[allow(unconditional_recursion)]
+fn stack_overflow() {
+    stack_overflow(); // 再帰のたびにリターンアドレスがプッシュされる
+    volatile::Volatile::new(0).read(); // 末尾最適化を防ぐ
+}
+```
+
+新しいGDTを初期化するために`gdt::init`関数を呼びます。`interrupts::init_idt`関数を呼び出す代わりに、すぐ後に説明する`init_test_idt`関数を呼びます。なぜなら、私達はパニックの代わりに`exit_qemu(QemuExitCode::Success)`をするカスタムしたダブルフォルトハンドラを登録したいからです。
+
+`stack_overflow`関数は`main.rs`の中にある関数とほとんど同じです。唯一の違いは関数の末尾で**[末尾呼び出し最適化]**と呼ばれるコンパイラの最適化を防ぐために[`Volativle`]タイプを使って追加の[volatile]読み込みを行っていることです。他のところでは、この最適化はコンパイラが最後の宣言が再帰関数呼び出しである関数を通常のループに変換することを許します。結果として、追加のスタックフレームが関数呼び出しではつくられず、スタックの使用量が変わらないままとなります。
```
Comment (JohnTitor) [inline blog/content/edition-2/posts/06-double-faults/index.ja.md:484]:
```suggestion
`stack_overflow`関数は`main.rs`の中にある関数とほとんど同じです。唯一の違いは[末尾呼び出し最適化]と呼ばれるコンパイラの最適化を防ぐために[`Volatile`]タイプを使って関数の末尾で追加の[volatile]読み込みを行っていることです。この最適化の特徴として、コンパイラが、最後の文が再帰関数呼び出しである関数を通常のループに変換できるようになる、というものがあります。その結果として、追加のスタックフレームが関数呼び出しではつくられず、スタックの使用量が変わらないままとなります。
```

File: blog/content/edition-2/posts/06-double-faults/index.ja.md [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+
+[volatile]: https://en.wikipedia.org/wiki/Volatile_(computer_programming)
+[`Volatile`]: https://docs.rs/volatile/0.2.6/volatile/struct.Volatile.html
+[末尾呼び出し最適化]: https://ja.wikipedia.org/wiki/末尾再帰#末尾呼出し最適化
+
+私達の場合は、しかしながら、スタックオーバーフローを起こしたいので、ダミーのコンパイラが除去することが許されていないvolatile読み込み文を関数の末尾に追加します。その結果、関数は決して**末尾再帰**ではなくなり、ループへの変換は防がれます。更に関数が無限に再帰することに対するコンパイラの警告をなくすために`allow(unconditional_recursion)`属性を追加します。
```
Comment (JohnTitor) [inline blog/content/edition-2/posts/06-double-faults/index.ja.md:490]:
```suggestion
しかし、ここではスタックオーバーフローを起こしたいので、コンパイラに削除されない、ダミーのvolatile読み込み文を関数の末尾に追加します。その結果、関数は**末尾再帰**ではなくなり、ループへの変換は防がれます。更に関数が無限に再帰することに対するコンパイラの警告をなくすために`allow(unconditional_recursion)`属性を追加します。
```

File: blog/content/edition-2/posts/06-double-faults/index.ja.md [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+
+### IDTのテスト 
+
+上で述べたように、テストはカスタムしたダブルフォルトハンドラを含む専用のIDTが必要です。実装はこのようになります。
```
Comment (JohnTitor) [inline blog/content/edition-2/posts/06-double-faults/index.ja.md:494]:
```suggestion
上で述べたように、テストはカスタムしたダブルフォルトハンドラを含む専用のIDTが必要です。実装はこのようになります：
```

File: blog/content/edition-2/posts/06-double-faults/index.ja.md [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+pub fn init_test_idt() {
+    TEST_IDT.load();
+}
+```
+
+実装は`interrupts.rs`内の通常のIDTと非常に似ています。通常のIDT同様、分離されたスタックに切り替えるようダブルフォルトハンドラ用のISTにスタックインデックスをセットします。`init_test_idt`関数は`load`メソッドによりCPU上にIDTを読み込みます。
+
+### ダブルフォルトハンドラ
+
+唯一欠けているのはダブルフォルトハンドラです。このようになります。
```
Comment (JohnTitor) [inline blog/content/edition-2/posts/06-double-faults/index.ja.md:524]:
```suggestion
唯一欠けているのはダブルフォルトハンドラです。このようになります：
```

File: blog/content/edition-2/posts/06-double-faults/index.ja.md [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+    _stack_frame: &mut InterruptStackFrame,
+    _error_code: u64,
+) -> ! {
+    serial_println!("[ok]");
+    exit_qemu(QemuExitCode::Success);
+    loop {}
+}
+```
+
+ダブルフォルトハンドラが呼ばれるとき、私達はQEMUを正常な終了コードで終了し、テストを成功とマークします。インテグレーションテストは完全に分けられた実行ファイルなので、私達はテストファイルの先頭で`#![feature(abi_x86_interrupt)]`属性を再びセットする必要があります。
```
Comment (JohnTitor) [inline blog/content/edition-2/posts/06-double-faults/index.ja.md:542]:
```suggestion
ダブルフォルトハンドラが呼ばれるとき、私達はQEMUを正常な終了コードで終了し、テストを成功とマークします。結合テストは完全に分けられた実行ファイルなので、私達はテストファイルの先頭で`#![feature(abi_x86_interrupt)]`属性を再びセットする必要があります。
```

File: blog/content/edition-2/posts/06-double-faults/index.ja.md [hunk 7]
```diff
@@ -0,0 +1,554 @@
...
+
+これで私達は`cargo test --test stack_overflow`（もしくは全部のテストを走らせるよう`cargo test`）でテストを走らせることができます。期待していたとおり、`stack_overflow... [ok]`とコンソールに出力されるのがわかります。`set_stack_index`の行をコメントアウトすると、テストは失敗するでしょう。
+
+## まとめ
+この記事では私達はダブルフォルトが何であるかとどういう条件下で発生するかを学びました。エラーメッセージを出力する基本的なダブルフォルトハンドラを追加しまし、そのためのインテグレーションテストを追加しました。
```
Comment (JohnTitor) [inline blog/content/edition-2/posts/06-double-faults/index.ja.md:547]:
```suggestion
この記事では私達はダブルフォルトが何であるかとどういう条件下で発生するかを学びました。エラーメッセージを出力する基本的なダブルフォルトハンドラと、そのための結合テストを追加しました。
```

File: blog/content/edition-2/posts/06-double-faults/index.ja.md [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+
+また、私達はスタックオーバーフローでも動くよう、ハードウェア支援によるダブルフォルト発生時のスタック切り替えを有効化しました。実装していく中で、古いアーキテクチャでのセグメンテーションで使われていたタスクステートセグメント（TSS）、割り込みスタックテーブル（IST）、グローバルディスクリプタテーブル（GDT）についても学びました。
```
Comment (JohnTitor) [inline blog/content/edition-2/posts/06-double-faults/index.ja.md:549]:
```suggestion
また、私達はスタックオーバーフロー下でも動くよう、ダブルフォルト発生時にハードウェアがサポートするスタック切り替えを行うようにしました。実装していく中で、古いアーキテクチャでのセグメンテーションで使われていたタスクステートセグメント（TSS）、割り込みスタックテーブル（IST）、グローバルディスクリプタテーブル（GDT）についても学びました。
```

File: blog/content/edition-2/posts/06-double-faults/index.ja.md [hunk 9]
```diff
@@ -0,0 +1,554 @@
...
+
+## 次は？
+次の記事ではどのようにタイマーやキーボードやネットワークコントローラのような外部デバイスからの割り込みを処理するかを説明します。これらのハードウェア割り込みは例外によく似ています。例えば、これらはIDTからディスパッチされます。しかしながら、例外とは違い、それらはCPUから直接発生しません。代わりに、**割り込みコントローラ**がこれらの割り込みを集めて、優先度によってそれらをCPUに向かわせます。次回は私達は[Intel 8259]（PIC）割り込みコントローラを調べ、どのようにキーボードのサポートを実装するかを学びます。
```
Comment (JohnTitor) [inline blog/content/edition-2/posts/06-double-faults/index.ja.md:552]:
```suggestion
次の記事ではタイマーやキーボード、ネットワークコントローラのような、外部デバイスからの割り込みをどのように処理するかを説明します。これらのハードウェア割り込みは例外によく似ています。例えば、これらもIDTからディスパッチされます。しかしながら、例外とは違い、それらはCPU上で直接発生するものではありません。代わりに、**割り込みコントローラ**がこれらの割り込みを集めて、優先度によってそれらをCPUに送ります。次回、私達は[Intel 8259]（PIC）割り込みコントローラを調べ、どのようにキーボードのサポートを実装するかを学びます。
```

**Review by JohnTitor — COMMENTED** [review]
Thanks for translating and sorry for the delay! Left some reviews roughly but almost looks great to me :)

**Comment by JohnTitor** [comment]
Btw, IIRC "Solve #XXX" won't close the issue, you should use "resolve", "close", or "fix" instead.

File: blog/content/edition-2/posts/06-double-faults/index.ja.md (hunk 2 above)
Comment (garasubo) [reply blog/content/edition-2/posts/06-double-faults/index.ja.md:484]: Any reason to remove bold style for 「末尾呼び出し最適化」？

**Review by JohnTitor — APPROVED** [review]
Thanks @garasubo!

@woodyZootopia I think it's good to go ahead as-is and follow-up stuff when you get some time to review, does it work for you?

**Comment by woodyZootopia** [comment]
Yes, thank you! I'm still not able to check the translation, so please proceed! I will create another PR/Issue if I were to find any problem.

**Comment by phil-opp** [comment]
Ok great, thanks a lot!

**Pull request merged**